src/net/server/handlers/create_query.rs

```diff
@@ -0,0 +1,246 @@
+    fn take_bytes(size_in_bytes: u32, buffered: &mut Vec<Bytes>, buffered_size: &mut u32) -> Bytes {
+        assert!(*buffered_size >= size_in_bytes);
+
+        // take from single buffer
+        // simply splits the first buffer in-place
+        if buffered[0].len() > size_in_bytes as usize {
+            let out_bytes = buffered[0].split_to(size_in_bytes as usize);
+            *buffered_size -= size_in_bytes;
+            out_bytes
+        } else {
+            // take across buffers
+            // removes expended buffers after taking
+            let mut out_bytes = Vec::new();
+            let mut remaining_bytes = size_in_bytes as usize;
+            let mut i = 0;
+            loop {
+                if remaining_bytes == 0 {
+                    break;
+                }
+                let buffer = &mut buffered[i];
+                let buffer_len = buffer.len();
+                // current `buffer` has more bytes than needed
+                if buffer_len > remaining_bytes {
+                    out_bytes.push(buffer.split_to(remaining_bytes));
+                    remaining_bytes = 0;
+                } else {
+                    // current `buffer` has <= bytes needed, needs removal
+                    out_bytes.push(buffer.clone());
+                    i += 1;
+                    remaining_bytes -= buffer.len();
+                }
+            }
+
+            // remove expended buffers
+            buffered.drain(0..i);
+            // reduce size of total buffers accordingly
+            *buffered_size -= size_in_bytes;
+            // create new `Bytes` that combines chunks
+            let bytes_vec = out_bytes.into_iter().fold(
+                Vec::with_capacity(size_in_bytes as usize),
+                |mut acc, n| {
+                    acc.extend_from_slice(&n);
+                    acc
+                },
+            );
+            Bytes::from(bytes_vec)
+        }
+    }
+}
+
+impl Stream for ByteArrStream {
+    type Item = Result<Bytes, MpcHelperServerError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut().project();
+        loop {
+            // if we currently have enough bytes, return them
+            if *this.buffered_size >= *this.size_in_bytes {
+                return Poll::Ready(Some(Ok(Self::take_bytes(
+                    *this.size_in_bytes,
+                    this.buffered,
+                    this.buffered_size,
+                ))));
+            }
+            // if we need more bytes, poll the body
+            #[allow(clippy::cast_possible_truncation)] // bytes size will not exceed u32
+            match ready!(this.body.as_mut().poll_next(cx)) {
+                // if body is expended, but we have some bytes leftover, error
+                None if *this.buffered_size > 0 => {
+                    return Poll::Ready(Some(Err(MpcHelperServerError::WrongBodyLen {
```

Comment:
this is a generic struct, returning mpchelper errors for it looks suspicious.

Reply:
ok. i can probably use some io errors instead